pkg/filters/validator/basicauth.go

```diff
@@ -313,6 +337,62 @@ func (euc *etcdUserCache) Match(username string, password string) bool {
 	return euc.userFileObject.Match(username, password)
 }
 
+func newLDAPUserCache(spec *ldapSpec) *ldapUserCache {
+	if spec.CertBase64 != "" && spec.KeyBase64 != "" {
+		certPem, _ := base64.StdEncoding.DecodeString(spec.CertBase64)
+		keyPem, _ := base64.StdEncoding.DecodeString(spec.KeyBase64)
+		if cert, err := tls.X509KeyPair(certPem, keyPem); err == nil {
+			spec.certificates = append(spec.certificates, cert)
+		} else {
+			logger.Errorf("generates x509 key pair failed: %v", err)
+		}
+	}
+	client := &ldap.LDAPClient{
+		Host:               spec.Host,
+		Port:               spec.Port,
+		Base:               spec.BaseDN,
+		UseSSL:             spec.UseSSL,
+		SkipTLS:            spec.SkipTLS,
+		InsecureSkipVerify: spec.Insecure,
+		ServerName:         spec.ServerName,
+		ClientCertificates: spec.certificates,
+	}
+	return &ldapUserCache{
+		spec:   spec,
+		client: client}
+}
+
+// make it mockable
+var fnAuthLDAP = func(luc *ldapUserCache, username, password string) bool {
+	defer luc.client.Close()
+	if err := luc.client.Connect(); err != nil {
+		logger.Warnf("failed to connect LDAP server %v", err)
+		return false
+	}
+
+	userdn := fmt.Sprintf("%s=%s,%s", luc.spec.UID, username, luc.spec.BaseDN)
+	if err := luc.client.Conn.Bind(userdn, password); err != nil {
+		logger.Warnf("failed to bind LDAP user %v", err)
+		return false
+	}
+
+	return true
```

Comment:
could this function be called concurrently?

Reply:
this quesion is very considerate, the `client.Connect()` will create a connection if not exists and then store it which can be used by other function like `Conn.Bind`, but the code `defer luc.client.Close()` will break it and could case concurrency problem, so I remove it.

Reply:
do we sure that `luc.client.Conn.Bind` is goroutine safe.

Reply:
it should be safe according to [https://github.com/go-ldap/ldap/blob/master/ldap_test.go#L198](https://github.com/go-ldap/ldap/blob/master/ldap_test.go#L198), and I also did a test which shows safe.